java-checks/src/main/java/org/sonar/java/checks/RegexPatternsNeedlesslyCheck.java

```diff
@@ -0,0 +1,101 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.regex.Pattern;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S4248")
+public class RegexPatternsNeedlesslyCheck extends AbstractMethodDetection {
+
+  private static final String STRING = "java.lang.String";
+  private static final String PATTERN = "java.util.regex.Pattern";
+  private static final Pattern SPLIT_EXCLUSION = Pattern.compile("[\\$\\.\\|\\(\\)\\[\\{\\^\\?\\*\\+\\\\]|\\\\\\w");
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      MethodMatcher.create().typeDefinition(PATTERN).name("compile").addParameter(STRING),
+      MethodMatcher.create().typeDefinition(STRING).name("matches").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(STRING).name("split").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(STRING).name("replaceAll").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(STRING).name("replaceFirst").withAnyParameters());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    ExpressionTree argument = mit.arguments().get(0);
+    if (mit.symbol().name().equals("split") && argument.is(Tree.Kind.STRING_LITERAL)) {
+      String argValue = LiteralUtils.trimQuotes(((LiteralTree) argument).value());
+      int strLength = argValue.length();
+      if (((strLength == 1 || (strLength == 2 && argValue.charAt(0) == '\\')) &&
+        !SPLIT_EXCLUSION.matcher(argValue).matches()) || metaCharactersInSplit(argValue)) {
+        return;
+      }
+    }
+    if (!storedInStaticFinal(mit) && (argument.is(Tree.Kind.STRING_LITERAL) || isConstant(argument))) {
+      reportIssue(mit, "Refactor this code to use a \"static final\" Pattern.");
+    }
+  }
+
+  private static boolean storedInStaticFinal(MethodInvocationTree mit) {
+    Tree tree = mit.parent();
+    while (!tree.is(Tree.Kind.VARIABLE, Tree.Kind.CLASS, Tree.Kind.ASSIGNMENT)) {
```

Comment:
You can probably stop at `METHOD` level as well. 
If you reach a `CLASS` or a `METHOD`, before an `ASSIGNMENT` or a `VARIABLE`, it can not be stored in a static final field.

Now, you may also want to group the `KIND`s you target by category for consistency (`CLASS` and `METHOD` are failing stop conditions, `ASSIGNMENT` and `VARIABLE` allows you to investigate further)